### DISCUSSION TITLE
Error: Empty installation

### DISCUSSION THREAD

**Discussion by asmaloney** [Homebrew/discussions#4144] [Everyday usage] [answered]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 3.6.20-143-g10845a1
ORIGIN: https://github.com/Homebrew/brew.git
HEAD: 10845a1122b916dc91d04f00a4d8b7b888d52654
Last commit: 7 hours ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 81f6a8a50fd95231d9bf177069596a5641d06598
Core tap last commit: 3 days ago
Core tap branch: master
HOMEBREW_PREFIX: /Users/foo/dev
HOMEBREW_REPOSITORY: /Users/foo/dev
HOMEBREW_CELLAR: /Users/foo/dev/Cellar
HOMEBREW_CASK_OPTS: []
HOMEBREW_GITHUB_API_TOKEN: set
HOMEBREW_INSTALL_FROM_API: set
HOMEBREW_MAKE_JOBS: 16
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: 16-core 64-bit kabylake
Clang: 13.1.6 build 1316
Git: 2.39.1 => /Users/foo/dev/bin/git
Curl: 7.79.1 => /usr/bin/curl
macOS: 12.6.2-x86_64
CLT: 14.2.0.0.1.1668646533
Xcode: 13.4.1
```

### Output of `brew doctor`

```shell
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Your Homebrew's prefix is not /usr/local.

Many of Homebrew's bottles (binary packages) can only be used with the default prefix.
Consider uninstalling Homebrew and reinstalling into the default prefix.
It is expected behaviour that some formulae will fail to build in this unsupported configuration.
It is expected behaviour that Homebrew will be buggy and slow.
Do not create any issues about this on Homebrew's GitHub repositories.
Do not create any issues even if you think this message is unrelated.
Any opened issues will be immediately closed without response.
Do not ask for help from MacHomebrew on Twitter.
You may ask for help in Homebrew's discussions but are unlikely to receive a response.
Try to figure out the problem yourself and submit a fix as a pull request.
We will review it but may or may not accept it.
```

### Description of issue

Recent updates download and build, but fail to install with a cryptic error message:

```
% brew upgrade imagemagick
==> Upgrading 1 outdated package:
imagemagick 7.1.0-58 -> 7.1.0-58_1
==> Fetching imagemagick
Warning: Building imagemagick from source as the bottle needs:
- HOMEBREW_CELLAR: /usr/local/Cellar (yours is /Users/foo/dev/Cellar)
- HOMEBREW_PREFIX: /usr/local (yours is /Users/foo/dev)
==> Downloading https://imagemagick.org/archive/releases/ImageMagick-7.1.0-58.tar.xz
Already downloaded: /Users/foo/Library/Caches/Homebrew/downloads/8373f899093049f659b8deccb8dc37eb7ac4037ee8a216a245e9efbfa957baf2--ImageMagick-7.1.0-58.tar.xz
==> Upgrading imagemagick
  7.1.0-58 -> 7.1.0-58_1

==> ./configure --prefix=/Users/foo/dev/Cellar/imagemagick/7.1.0-58 --libdir=/Users/foo/dev/Cellar/imagemagick/7.1.0-58/lib --enable-osx-universal-binary=no --disable-opencl --enable-shared --enable-st
==> make install
Error: Empty installation
```

The same thing happens with `node` (which is less surprising to me because it's a constant problem).

I can't find any info about this error - what does it mean? Other packages have installed OK.

**Comment by SMillerDev** [comment]
Have you tried building from source?

**Reply by asmaloney** [reply]
These are building from source:
```
==> ./configure --prefix=/Users/foo/dev/Cellar/imagemagick/7.1.0-58 --libdir=/Users/foo/dev/Cellar/imagemagick/7.1.0-58/lib --enable-osx-universal-binary=no --disable-opencl --enable-shared --enable-st
==> make install
```

Either that or they are taking a long time to do nothing 😆

**Reply by SMillerDev** [reply]
But are you specifically requesting a source install?

**Reply by asmaloney** [reply]
Oh - no. I've always (for many years) let homebrew manage what needs to be built and when.

This is new since last week(_edit_ -ish) - I wonder if it's related to 3.6.20?

Does requesting a source install do something different from homebrew forcing a source install (as above)?

(I also had to give up with node on my laptop because it ate through 13.6 GB of disk space and ran out of room. Oh, modern software 🙄 👴  )

**Reply by asmaloney** [reply]
It looks like it might be any packages that are trying to compile themselves since it just happened with `cppcheck`. When I try it with something bottled (like `bullet`) it works.

I just tried `cppcheck` on another machine (same homebrew, macOS version, and path) and it built and installed properly, so I guess there's something not linked properly or missing on my main machine. Any idea what `Error: Empty installation` means? It's happening after it's built properly when trying to install it.

I can't find any info on it.

**Comment by asmaloney** [comment]
I wonder if I'm seeing the same things as [homebrew-core #121566](https://github.com/Homebrew/homebrew-core/pull/121566)...

**Reply by gromgit** [reply]
`export HOMEBREW_NO_INSTALL_FROM_API=1` and try your source build again.

**Reply by frakman1** [reply]
This worked for me.
```bash
$export HOMEBREW_NO_INSTALL_FROM_API=1
$brew reinstall sqlite
```

**Comment by asmaloney** [comment, marked as answer]
@gromgit:
> `export HOMEBREW_NO_INSTALL_FROM_API=1` and try your source build again.

**Yes!** Thank you.

(Though I don't really understand why one of my machines needs this and the other doesn't since they are set up the same way...)

Are there any docs for this? The [PR comment](https://github.com/Homebrew/brew/pull/14375) just says that it disables the `HOMEBREW_INSTALL_FROM_API` functionality 😄 I can't find any user docs for that one.

**Reply by SMillerDev** [reply]
> (Though I don't really understand why one of my machines needs this and the other doesn't since they are set up the same way...)

What does `brew developer` say on both?

**Reply by asmaloney** [reply]
That's it! Another mystery solved. I must have turned that on at some point in the last 12-ish years for some reason 😆

Thank you @SMillerDev!

**Reply by asmaloney** [reply]
That might be a good thing to add to the `brew doctor` output.
